File: src-tauri/src/commands/commits.rs
use serde::Serialize;
use tauri::State;

use crate::{
    domain::Commit,
    infrastructure::{
        database::connection::get_connection,
        git::{CommitDiff, CommitInlineDiff, FileDiff},
    },
    services::commit_service,
    state::AppState,
};

#[derive(Debug, Serialize)]
pub struct CommitResponse {
    pub id: i64,
    pub hash: String,
    pub short_hash: String,
    pub author_name: String,
    pub author_email: String,
    pub committer_name: String,
    pub committer_email: String,
    pub subject: String,
    pub body: Option<String>,
    pub parent_count: u32,
    pub committed_at: String,
    pub is_significant: bool,
    pub is_merge_commit: bool,
    pub is_root_commit: bool,
}

impl From<Commit> for CommitResponse {
    fn from(c: Commit) -> Self {
        let merge = c.is_merge_commit();
        let root = c.is_root_commit();
        Self {
            short_hash: c.hash.short(),
            id: c.id.0,
            hash: c.hash.full().to_string(),
            author_name: c.author_name,
            author_email: c.author_email,
            committer_name: c.committer_name,
            committer_email: c.committer_email,
            subject: c.subject,
            body: c.body,
            parent_count: c.parent_count,
            committed_at: c.committed_at,
            is_significant: c.is_significant,
            is_merge_commit: merge,
            is_root_commit: root,
        }
    }
}

#[tauri::command]
pub fn get_commits(
    repo_id: i64,
    count: Option<usize>,
    offset: Option<usize>,
    state: State<'_, AppState>,
) -> Result<Vec<CommitResponse>, String> {
    let conn = get_connection(&state.db_path).map_err(|e| e.to_string())?;
    commit_service::get_commits(&conn, repo_id, count.unwrap_or(50), offset.unwrap_or(0))
        .map(|commits| commits.into_iter().map(CommitResponse::from).collect())
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub fn get_commit_by_hash(
    repo_id: i64,
    hash: String,
    state: State<'_, AppState>,
) -> Result<CommitResponse, String> {
    let conn = get_connection(&state.db_path).map_err(|e| e.to_string())?;
    commit_service::get_commit_by_hash(&conn, repo_id, &hash)
        .map(CommitResponse::from)
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub fn get_commit_diff(
    repo_id: i64,
    commit_hash: String,
    state: State<'_, AppState>,
) -> Result<CommitDiff, String> {
    let conn = get_connection(&state.db_path).map_err(|e| e.to_string())?;
    commit_service::get_commit_diff(&conn, repo_id, &commit_hash).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn get_commit_inline_diff(
    repo_id: i64,
    commit_hash: String,
    state: State<'_, AppState>,
) -> Result<CommitInlineDiff, String> {
    let conn = get_connection(&state.db_path).map_err(|e| e.to_string())?;
    commit_service::get_commit_inline_diff(&conn, repo_id, &commit_hash).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn get_file_diff(
    repo_id: i64,
    commit_hash: String,
    file_path: String,
    state: State<'_, AppState>,
) -> Result<FileDiff, String> {
    let conn = get_connection(&state.db_path).map_err(|e| e.to_string())?;
    commit_service::get_file_diff_by_commit_hash(&conn, repo_id, &commit_hash, &file_path)
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub fn get_file_diff_history(
    repo_id: i64,
    file_path: String,
    commit_count: Option<usize>,
    commit_offset: Option<usize>,
    state: State<'_, AppState>,
) -> Result<Vec<FileDiff>, String> {
    let conn = get_connection(&state.db_path).map_err(|e| e.to_string())?;
    commit_service::get_file_diff_history(
        &conn,
        repo_id,
        &file_path,
        commit_count.unwrap_or(10),
        commit_offset.unwrap_or(0),
    )
    .map_err(|e| e.to_string())
}

File: src/lib/tauri/commits.ts
import { tauri } from "./tauri";

export interface Commit {
  id: number,
  hash: string,
  shortHash: string,
  authorName: string,
  authorEmail: string,
  committerName: string,
  committerEmail: string,
  subject: string,
  body: string | null,
  parentCount: number,
  committedAt: string,
  isSignificant: boolean,
  isMergeCommit: boolean,
  isRootCommit: boolean,
}

interface CommitResponse {
  id: number,
  hash: string,
  short_hash: string,
  author_name: string,
  author_email: string,
  committer_name: string,
  committer_email: string,
  subject: string,
  body: string | null,
  parent_count: number,
  committed_at: string,
  is_significant: boolean,
  is_merge_commit: boolean,
  is_root_commit: boolean,
}

export interface CommitGraphNode {
  hash: string,
  branchName: string | null,
  branchNames: string[],
  authorName: string,
  authorEmail: string,
  subject: string,
  committedAt: string,
  additions: number,
  deletions: number,
  totalChangedFilesCount: number,
  parentHashes: string[],
}

interface CommitGraphNodeResponse {
  hash: string,
  branch_name: string | null,
  branch_names: string[],
  author_name: string,
  author_email: string,
  subject: string,
  committed_at: string,
  additions: number,
  deletions: number,
  total_changed_files_count: number,
  parent_hashes: string[],
}

function toCommit(commit: CommitResponse): Commit {
  return {
    id: commit.id,
    hash: commit.hash,
    shortHash: commit.short_hash,
    authorName: commit.author_name,
    authorEmail: commit.author_email,
    committerName: commit.committer_name,
    committerEmail: commit.committer_email,
    subject: commit.subject,
    body: commit.body,
    parentCount: commit.parent_count,
    committedAt: commit.committed_at,
    isSignificant: commit.is_significant,
    isMergeCommit: commit.is_merge_commit,
    isRootCommit: commit.is_root_commit,
  };
}

function toCommitGraphNode(node: CommitGraphNodeResponse): CommitGraphNode {
  return {
    hash: node.hash,
    branchName: node.branch_name,
    branchNames: node.branch_names,
    authorName: node.author_name,
    authorEmail: node.author_email,
    subject: node.subject,
    committedAt: node.committed_at,
    additions: node.additions,
    deletions: node.deletions,
    totalChangedFilesCount: node.total_changed_files_count,
    parentHashes: node.parent_hashes,
  };
}

export async function getCommitByHash(repoId: number, hash: string): Promise<Commit> {
  const commitResponse = await tauri<CommitResponse>('get_commit_by_hash', { repoId, hash });
  return toCommit(commitResponse);
}

export async function getCommitsByRepoId(repoId: number, limit: number, offset: number): Promise<Commit[]> {
  const commitsResponse = await tauri<CommitResponse[]>('get_commits', { repoId, count: limit, offset });
  return commitsResponse.map(toCommit);
}

export async function getCommitGraphByRepoId(repoId: number, limit: number, offset: number): Promise<CommitGraphNode[]> {
  const commitGraphResponse = await tauri<CommitGraphNodeResponse[]>('get_commit_graph', { repoId, count: limit, offset });
  return commitGraphResponse.map(toCommitGraphNode);
}

export async function getCommitDiffByHash(repoId: number, hash: string): Promise<any> {
  const diff = await tauri<any>('get_commit_diff', { repoId, commitHash: hash });
  return diff;
}

export async function getCommitInlineDiff(repoId: number, hash: string): Promise<any> {
  const inlineDiff = await tauri<any>('get_commit_inline_diff', { repoId, commitHash: hash });
  return inlineDiff;
}
